rust/sdk-processor/src/utils/starting_version.rs

```diff
@@ -1,52 +1,278 @@
 use super::database::ArcDbPool;
-use crate::config::indexer_processor_config::IndexerProcessorConfig;
+use crate::{
+    config::indexer_processor_config::IndexerProcessorConfig,
+    db::common::models::{
+        backfill_processor_status::{BackfillProcessorStatusQuery, BackfillStatus},
+        processor_status::ProcessorStatusQuery,
+    },
+};
 use anyhow::{Context, Result};
-use processor::db::common::models::processor_status::ProcessorStatusQuery;
 
+/// Get the appropriate starting version for the processor.
+///
+/// If it is a regular processor, this will return the higher of the checkpointed version,
+/// or `staring_version` from the config, or 0 if not set.
+///
+/// If this is a backfill processor and threre is an in-progress backfill, this will return
+/// the checkpointed version + 1.
+///
+/// If this is a backfill processor and there is not an in-progress backfill (i.e., no checkpoint or
+/// backfill status is COMPLETE), this will return `starting_version` from the config, or 0 if not set.
 pub async fn get_starting_version(
     indexer_processor_config: &IndexerProcessorConfig,
     conn_pool: ArcDbPool,
 ) -> Result<u64> {
-    // If starting_version is set in TransactionStreamConfig, use that
-    if indexer_processor_config
-        .transaction_stream_config
-        .starting_version
-        .is_some()
-    {
-        return Ok(indexer_processor_config
-            .transaction_stream_config
-            .starting_version
-            .unwrap());
-    }
-
-    // If it's not set, check if the DB has latest_processed_version set and use that
-    let latest_processed_version_from_db =
+    // Check if there's a checkpoint in the approrpiate processor status table.
+    let latest_processed_version =
         get_latest_processed_version_from_db(indexer_processor_config, conn_pool)
             .await
             .context("Failed to get latest processed version from DB")?;
-    if let Some(latest_processed_version_tracker) = latest_processed_version_from_db {
-        // Add +1 to start from the version after the last successful version
-        return Ok(latest_processed_version_tracker + 1);
-    }
 
-    // If latest_processed_version is not stored in DB, return the default 0
-    Ok(0)
+    // If nothing checkpointed, return the `starting_version` from the config, or 0 if not set.
+    Ok(latest_processed_version.unwrap_or(
+        indexer_processor_config
+            .transaction_stream_config
+            .starting_version
+            .unwrap_or(0),
+    ))
```

Comment:
This changes the existing priority of the flags for regular processors which can make it hard if you're developing locally and want to start from a specific version. Could we keep the order for regular processor?

Reply:
you mean the use case where you want `starting_version` to have higher precedence than last_processed_version? 

That's a trade-off we have to make to have better restart behavior. Otherwise, the regular processor will always restart at the starting_version if that's ever set

Reply:
I could add the old `starting_version_override` config option lol. But rename it to `starting_version_dev_override` to make it clear its for development purposes so its easier for local development. Wdyt?

Reply:
What if we keep the priority the same for regular processors to match with existing devex, and establish a different priority only for backfill processors?

Reply:
> head processors in prod should never need to use starting_version in the config

how come? if we made a logic change, wouldn't the steps be something like
1. create a backfill processor from 0 to HEAD
2. create a regular processor from HEAD onwards

to do #2, wouldn't we use `starting_version=HEAD` where HEAD is just some recent version?

Reply:
Good point! I was thinking about the current processors already running in prod, but if you're a new processor and you set `starting_version=HEAD` then going forward, it should only need to refer to the latest version from DB. 

And to my point of developing locally, they can also use the `backfill_processor` config to start from arbitrary version!

Reply:
awesome. thanks for the discussion! yeah, or even run diesel reset to wipe the state